// quais/lib/esm/wallet/qi-wallets/bip47-paymentchannel.js
import { Bip47QiWalletSelf } from './bip47-self-qi-wallet.js';
import { Bip47QiWalletCounterparty } from './bip47-counterparty-qi-wallet.js';
/**
 * Represents a BIP47 payment channel between two parties, managing both sending and receiving addresses for a specific
 * payment code pair.
 */
export class PaymentChannel {
    #selfWallet;
    #counterpartyWallet;
    #counterpartyPaymentCode;
    constructor(root, counterpartyPaymentCode) {
        this.#counterpartyPaymentCode = counterpartyPaymentCode;
        this.#selfWallet = new Bip47QiWalletSelf(root, counterpartyPaymentCode);
        this.#counterpartyWallet = new Bip47QiWalletCounterparty(this.#selfWallet, counterpartyPaymentCode);
    }
    /**
     * Gets the self wallet instance used for receiving payments
     */
    get selfWallet() {
        return this.#selfWallet;
    }
    /**
     * Gets the counterparty wallet instance used for sending payments
     */
    get counterpartyWallet() {
        return this.#counterpartyWallet;
    }
    /**
     * Gets the counterparty's payment code
     */
    get counterpartyPaymentCode() {
        return this.#counterpartyPaymentCode;
    }
    /**
     * Gets the address info for a given receiving address
     *
     * @param address - The receiving address to get the info for
     * @returns The address info or null if not found
     */
    getReceivingAddressInfo(address) {
        return this.#selfWallet.getAddressInfo(address);
    }
    getReceivingAddressesForAccount(account) {
        return this.#selfWallet.getAddressesForAccount(account);
    }
    getNextReceivingAddress(zone, account = 0) {
        return this.#selfWallet.deriveNewAddress(zone, account);
    }
    getNextSendingAddress(zone, account = 0) {
        return this.#counterpartyWallet.deriveNewAddress(zone, account);
    }
}
