// quais/lib/commonjs/wallet/qi-wallets/bip47-paymentchannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentChannel = void 0;
const bip47_self_qi_wallet_js_1 = require("./bip47-self-qi-wallet.js");
const bip47_counterparty_qi_wallet_js_1 = require("./bip47-counterparty-qi-wallet.js");
/**
 * Represents a BIP47 payment channel between two parties, managing both sending and receiving addresses for a specific
 * payment code pair.
 */
class PaymentChannel {
    #selfWallet;
    #counterpartyWallet;
    #counterpartyPaymentCode;
    constructor(root, counterpartyPaymentCode) {
        this.#counterpartyPaymentCode = counterpartyPaymentCode;
        this.#selfWallet = new bip47_self_qi_wallet_js_1.Bip47QiWalletSelf(root, counterpartyPaymentCode);
        this.#counterpartyWallet = new bip47_counterparty_qi_wallet_js_1.Bip47QiWalletCounterparty(this.#selfWallet, counterpartyPaymentCode);
    }
    /**
     * Gets the self wallet instance used for receiving payments
     */
    get selfWallet() {
        return this.#selfWallet;
    }
    /**
     * Gets the counterparty wallet instance used for sending payments
     */
    get counterpartyWallet() {
        return this.#counterpartyWallet;
    }
    /**
     * Gets the counterparty's payment code
     */
    get counterpartyPaymentCode() {
        return this.#counterpartyPaymentCode;
    }
    /**
     * Gets the address info for a given receiving address
     *
     * @param address - The receiving address to get the info for
     * @returns The address info or null if not found
     */
    getReceivingAddressInfo(address) {
        return this.#selfWallet.getAddressInfo(address);
    }
    getReceivingAddressesForAccount(account) {
        return this.#selfWallet.getAddressesForAccount(account);
    }
    getNextReceivingAddress(zone, account = 0) {
        return this.#selfWallet.deriveNewAddress(zone, account);
    }
    getNextSendingAddress(zone, account = 0) {
        return this.#counterpartyWallet.deriveNewAddress(zone, account);
    }
}
exports.PaymentChannel = PaymentChannel;
