// quais/lib/commonjs/wallet/qi-wallets/bip47-counterparty-qi-wallet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bip47QiWalletCounterparty = void 0;
const abstract_qi_wallet_js_1 = require("./abstract-qi-wallet.js");
const qi_hdwallet_js_1 = require("../qi-hdwallet.js");
const bip32_js_1 = require("../bip32/bip32.js");
const secp256k1_1 = __importDefault(require("@bitcoinerlab/secp256k1"));
const index_js_1 = require("../../utils/index.js");
const crypto_js_1 = require("../bip32/crypto.js");
const payment_codes_js_1 = require("../payment-codes.js");
const utils_js_1 = require("../utils.js");
const utils_js_2 = require("../utils.js");
class Bip47QiWalletCounterparty extends abstract_qi_wallet_js_1.AbstractQiWallet {
    selfBip47Wallet;
    counterpartyPaymentCode;
    static DEFAULT_GAP_LIMIT = 5;
    constructor(selfBip47Wallet, counterpartyPaymentCode) {
        super(Bip47QiWalletCounterparty.DEFAULT_GAP_LIMIT);
        this.selfBip47Wallet = selfBip47Wallet;
        this.counterpartyPaymentCode = counterpartyPaymentCode;
    }
    /**
     * Derives a new BIP47 payment address for sending funds to a counterparty.
     *
     * @remarks
     * This method implements BIP-0047 payment code derivation for sending funds to a counterparty. It derives addresses
     * using the counterparty's payment code until finding one that matches the specified zone. The derivation follows
     * the path specified in BIP-0047 using the next unused index.
     * @param {Zone} zone - The zone to derive the address for
     * @param {number} [account=0] - The account index to use for derivation. Default is `0`
     * @param {string} receiverPaymentCode - The Base58Check encoded payment code of the receiver
     * @returns {QiAddressInfo} Information about the newly derived payment address
     * @throws {Error} If the payment code version is invalid
     * @throws {Error} If unable to derive a valid address for the zone after maximum attempts
     */
    deriveNewAddress(zone, account = 0) {
        const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_1.default);
        const buf = crypto_js_1.bs58check.decode(this.counterpartyPaymentCode);
        const version = buf[0];
        if (version !== payment_codes_js_1.PC_VERSION)
            throw new Error('Invalid payment code version');
        const walletPCodePrivate = this.selfBip47Wallet.getPaymentCodePrivate(account);
        const receiverPCodePublic = new payment_codes_js_1.PaymentCodePublic(secp256k1_1.default, bip32, buf.slice(1));
        const lastIndex = this.getLastDerivationIndex(zone, account);
        let addrIndex = lastIndex + 1;
        for (let attempts = 0; attempts < utils_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS; attempts++) {
            const address = receiverPCodePublic.getPaymentAddress(walletPCodePrivate, addrIndex);
            if ((0, utils_js_2.isValidAddressForZone)(this.coinType, address, zone)) {
                this.saveLastDerivationIndex(zone, account, addrIndex);
                const pubkey = receiverPCodePublic.derivePaymentPublicKey(walletPCodePrivate, addrIndex);
                const pcInfo = {
                    address,
                    pubKey: (0, index_js_1.hexlify)(pubkey),
                    index: addrIndex,
                    account,
                    zone,
                    change: false,
                    status: qi_hdwallet_js_1.AddressStatus.UNKNOWN,
                    derivationPath: this.counterpartyPaymentCode,
                    lastSyncedBlock: null,
                };
                this.saveQiAddressInfo(pcInfo);
                return pcInfo;
            }
            addrIndex++;
        }
        throw new Error(`Failed to derive a valid address for the zone ${zone} after ${utils_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS} attempts.`);
    }
}
exports.Bip47QiWalletCounterparty = Bip47QiWalletCounterparty;
