// quais/lib/commonjs/wallet/qi-wallets/bip47-self-qi-wallet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bip47QiWalletSelf = exports.getPrivateKeyFromPaymentCode = exports.generatePaymentCodePrivate = void 0;
const abstract_qi_wallet_js_1 = require("./abstract-qi-wallet.js");
const qi_hdwallet_js_1 = require("../qi-hdwallet.js");
const bip32_js_1 = require("../bip32/bip32.js");
const secp256k1_1 = __importDefault(require("@bitcoinerlab/secp256k1"));
const types_js_1 = require("../bip32/types.js");
const index_js_1 = require("../../utils/index.js");
const crypto_js_1 = require("../bip32/crypto.js");
const payment_codes_js_1 = require("../payment-codes.js");
const utils_js_1 = require("../utils.js");
const utils_js_2 = require("../utils.js");
/**
 * Generates a BIP47 private payment code for the specified account. The payment code is created by combining the
 * account's public key and chain code.
 *
 * @param {HDNodeWallet} root - The root HD node to derive from
 * @param {number} account - The account index for which to generate the private payment code.
 * @returns {PaymentCodePrivate} The PaymentCodePrivate instance.
 */
function generatePaymentCodePrivate(root, account) {
    const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_1.default);
    const accountNode = root.deriveChild(account + utils_js_1.HARDENED_OFFSET);
    // payment code array
    const pc = new Uint8Array(80);
    // set version + options
    pc.set([1, 0]);
    // set the public key
    const pubKey = accountNode.publicKey;
    pc.set((0, index_js_1.getBytes)(pubKey), 2);
    // set the chain code
    const chainCode = accountNode.chainCode;
    pc.set((0, index_js_1.getBytes)(chainCode), 35);
    const adapter = new types_js_1.HDNodeBIP32Adapter(accountNode);
    return new payment_codes_js_1.PaymentCodePrivate(adapter, secp256k1_1.default, bip32, pc);
}
exports.generatePaymentCodePrivate = generatePaymentCodePrivate;
/**
 * Derives a private key from a BIP47 payment code for a specific index and account. This function implements the BIP47
 * specification for deriving private keys from payment codes.
 *
 * @param {HDNodeWallet} root - The root HD node to derive from
 * @param {string} counterPartyPaymentCode - The Base58Check encoded payment code of the counterparty
 * @param {number} index - The index to use for key derivation
 * @param {number} account - The account index to use for derivation
 * @returns {string} The derived private key as a hex string
 * @throws {Error} If the payment code version is invalid
 */
function getPrivateKeyFromPaymentCode(root, counterPartyPaymentCode, index, account) {
    const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_1.default);
    const buf = crypto_js_1.bs58check.decode(counterPartyPaymentCode);
    const version = buf[0];
    if (version !== payment_codes_js_1.PC_VERSION)
        throw new Error('Invalid payment code version');
    const counterpartyPCodePublic = new payment_codes_js_1.PaymentCodePublic(secp256k1_1.default, bip32, buf.slice(1));
    const paymentCodePrivate = generatePaymentCodePrivate(root, account);
    const paymentPrivateKey = paymentCodePrivate.derivePaymentPrivateKey(counterpartyPCodePublic, index);
    return (0, index_js_1.hexlify)(paymentPrivateKey);
}
exports.getPrivateKeyFromPaymentCode = getPrivateKeyFromPaymentCode;
class Bip47QiWalletSelf extends abstract_qi_wallet_js_1.AbstractQiWallet {
    root;
    counterpartyPaymentCode;
    static DEFAULT_GAP_LIMIT = 5;
    constructor(root, counterpartyPaymentCode) {
        super(Bip47QiWalletSelf.DEFAULT_GAP_LIMIT);
        this.root = root;
        this.counterpartyPaymentCode = counterpartyPaymentCode;
    }
    /**
     * Generates a BIP47 private payment code for the specified account. The payment code is created by combining the
     * account's public key and chain code.
     *
     * @private
     * @param {number} account - The account index for which to generate the private payment code.
     * @returns {Promise<PaymentCodePrivate>} A promise that resolves to the PaymentCodePrivate instance.
     */
    getPaymentCodePrivate(account = 0) {
        return generatePaymentCodePrivate(this.root, account);
    }
    /**
     * Derives a new BIP47 receiving address for a specific zone and account using the sender's payment code. This
     * method follows the BIP47 specification for deriving payment addresses from a notification transaction.
     *
     * @param {Zone} zone - The zone to derive the address for
     * @param {number} [account=0] - The account index to use for derivation. Default is `0`
     * @param {string} senderPaymentCode - The Base58Check encoded payment code of the sender
     * @returns {QiAddressInfo} Information about the newly derived address including:
     *
     *   - The derived address
     *   - Public key
     *   - Derivation index
     *   - Account number
     *   - Zone
     *   - Status and other metadata
     *
     * @throws {Error} If the payment code version is invalid
     * @throws {Error} If unable to derive a valid address for the zone after maximum attempts
     */
    deriveNewAddress(zone, account = 0) {
        const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_1.default);
        const buf = crypto_js_1.bs58check.decode(this.counterpartyPaymentCode);
        const version = buf[0];
        if (version !== payment_codes_js_1.PC_VERSION)
            throw new Error('Invalid payment code version');
        const senderPCodePublic = new payment_codes_js_1.PaymentCodePublic(secp256k1_1.default, bip32, buf.slice(1));
        const walletPCodePrivate = this.getPaymentCodePrivate(account);
        const lastIndex = this.getLastDerivationIndex(zone, account);
        let addrIndex = lastIndex + 1;
        for (let attempts = 0; attempts < utils_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS; attempts++) {
            const address = walletPCodePrivate.getPaymentAddress(senderPCodePublic, addrIndex);
            if ((0, utils_js_2.isValidAddressForZone)(this.coinType, address, zone)) {
                this.saveLastDerivationIndex(zone, account, addrIndex);
                const pubkey = walletPCodePrivate.derivePaymentPublicKey(senderPCodePublic, addrIndex);
                const pcInfo = {
                    address,
                    pubKey: (0, index_js_1.hexlify)(pubkey),
                    index: addrIndex,
                    account,
                    zone,
                    change: false,
                    status: qi_hdwallet_js_1.AddressStatus.UNKNOWN,
                    derivationPath: this.counterpartyPaymentCode,
                    lastSyncedBlock: null,
                };
                this.saveQiAddressInfo(pcInfo);
                return pcInfo;
            }
            addrIndex++;
        }
        throw new Error(`Failed to derive a valid address for the zone ${zone} after ${utils_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS} attempts.`);
    }
}
exports.Bip47QiWalletSelf = Bip47QiWalletSelf;
