// quais/lib/commonjs/wallet/qi-wallets/privkey-qi-wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivatekeyQiWallet = void 0;
const abstract_qi_wallet_js_1 = require("./abstract-qi-wallet.js");
const index_js_1 = require("../../address/index.js");
const index_js_2 = require("../../address/index.js");
const qi_hdwallet_js_1 = require("../qi-hdwallet.js");
const index_js_3 = require("../../utils/index.js");
const index_js_4 = require("../../crypto/index.js");
class PrivatekeyQiWallet extends abstract_qi_wallet_js_1.AbstractQiWallet {
    static DEFAULT_GAP_LIMIT = 0;
    constructor() {
        super(PrivatekeyQiWallet.DEFAULT_GAP_LIMIT);
    }
    /**
     * Imports a private key and adds it to the wallet.
     *
     * @param {string} privateKey - The private key to import (hex string)
     * @returns {Promise<QiAddressInfo>} The address information for the imported key
     * @throws {Error} If the private key or derived address is invalid
     */
    importPrivateKey(privateKey) {
        if (!(0, index_js_3.isHexString)(privateKey, 32)) {
            throw new Error(`Invalid private key format: must be 32-byte hex string (got ${privateKey})`);
        }
        const pubKey = index_js_4.SigningKey.computePublicKey(privateKey, true);
        const address = (0, index_js_2.computeAddress)(pubKey);
        // Validate address is for correct zone and ledger
        const addressZone = (0, index_js_3.getZoneForAddress)(address);
        if (!addressZone) {
            throw new Error(`Private key does not correspond to a valid address for any zone (got ${address})`);
        }
        if (!(0, index_js_1.isQiAddress)(address)) {
            throw new Error(`Private key does not correspond to a valid Qi address (got ${address})`);
        }
        if (this.addresses.has(address)) {
            throw new Error(`Address ${address} already exists in wallet`);
        }
        const addressInfo = {
            pubKey,
            address,
            account: 0,
            index: -1,
            change: false,
            zone: addressZone,
            status: qi_hdwallet_js_1.AddressStatus.UNUSED,
            derivationPath: privateKey,
            lastSyncedBlock: null,
        };
        this.addresses.set(address, addressInfo);
        return addressInfo;
    }
    getImportedAddresses(zone) {
        if (zone !== undefined) {
            return [...this.addresses.values()].filter((info) => info.zone === zone);
        }
        return [...this.addresses.values()];
    }
    getAddressesForAccount(account) {
        return [...this.addresses.values()].filter((info) => info.account === account);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    deriveNewAddress(zone, account = 0) {
        throw new Error('PrivatekeyQiWallet does not support deriveNewAddress');
    }
}
exports.PrivatekeyQiWallet = PrivatekeyQiWallet;
