// quais/lib/esm/wallet/abstract-hdwallet.js
import { HDNodeWallet } from './hdnodewallet.js';
import { Mnemonic } from './mnemonic.js';
import { LangEn } from '../wordlists/lang-en.js';
import { randomBytes } from '../crypto/index.js';
import { assertPrivate } from '../utils/index.js';
import { Zone } from '../constants/index.js';
export const _guard = {};
/**
 * Abstract class representing a Hierarchical Deterministic (HD) wallet.
 */
class AbstractHDWallet {
    static _version = 1;
    static _coinType;
    /**
     * Root node of the HD wallet.
     */
    _root;
    // Map of addresses to address info
    _addresses = new Map();
    provider;
    /**
     * @param {HDNodeWallet} root - The root node of the HD wallet.
     * @param {Provider} [provider] - The provider for the HD wallet.
     */
    constructor(guard, root, provider) {
        assertPrivate(guard, _guard, 'AbstractHDWallet');
        this._root = root;
        this.provider = provider;
    }
    /**
     * Returns the parent path for a given coin type.
     *
     * @param {number} coinType - The coin type.
     * @returns {string} The parent path.
     */
    static parentPath(coinType) {
        return `m/44'/${coinType}'`;
    }
    /**
     * Returns the coin type of the wallet.
     *
     * @returns {AllowedCoinType} The coin type.
     */
    coinType() {
        return this.constructor._coinType;
    }
    /**
     * Creates an instance of the HD wallet.
     *
     * @param {new (root: HDNodeWallet) => T} this - The constructor of the HD wallet.
     * @param {Mnemonic} mnemonic - The mnemonic.
     * @returns {T} The created instance.
     */
    static createInstance(mnemonic) {
        const coinType = this._coinType;
        const path = this.parentPath(coinType);
        const root = HDNodeWallet.fromMnemonic(mnemonic, path);
        return new this(_guard, root);
    }
    /**
     * Creates an HD wallet from a mnemonic.
     *
     * @param {new (root: HDNodeWallet) => T} this - The constructor of the HD wallet.
     * @param {Mnemonic} mnemonic - The mnemonic.
     * @returns {T} The created instance.
     */
    static fromMnemonic(mnemonic) {
        return this.createInstance(mnemonic);
    }
    static async fromMnemonicReactNative(mnemonic) {
        const coinType = this._coinType;
        const path = this.parentPath(coinType);
        const root = await HDNodeWallet.fromMnemonicReactNative(mnemonic, path);
        return new this(_guard, root);
    }
    /**
     * Creates an instance of the HD wallet from a root HD node.
     *
     * This method creates a wallet instance directly from an existing HD node root, optionally passing the original
     * seed for wallets that need it for additional derivation paths (like BIP47 payment codes).
     *
     * @param {new (guard: any, root: HDNodeWallet, seed?: BytesLike) => T} this - The constructor of the HD wallet.
     * @param {HDNodeWallet} root - The root HD node to use for the wallet.
     * @param {BytesLike} [seed] - Optional original seed bytes, needed for some wallet types.
     * @returns {T} The created wallet instance.
     * @protected
     */
    static createInstanceFromRoot(root, seed) {
        return new this(_guard, root, seed);
    }
    /**
     * Creates an HD wallet from a seed.
     *
     * This method creates a wallet by first generating an HD node from the provided seed, then deriving the appropriate
     * path based on the coin type, and finally creating a wallet instance with the derived root node.
     *
     * @param {new (guard: any, root: HDNodeWallet) => T} this - The constructor of the HD wallet.
     * @param {BytesLike} seed - The seed bytes used to generate the wallet.
     * @returns {T} The created wallet instance.
     */
    static fromSeed(seed) {
        let root = HDNodeWallet.fromSeed(seed);
        const coinType = this._coinType;
        const path = this.parentPath(coinType);
        root = root.derivePath(path);
        return this.createInstanceFromRoot(root, seed);
    }
    /**
     * Creates a random HD wallet.
     *
     * @param {new (root: HDNodeWallet) => T} this - The constructor of the HD wallet.
     * @param {string} [password] - The password.
     * @param {Wordlist} [wordlist] - The wordlist.
     * @returns {T} The created instance.
     */
    static createRandom(password, wordlist) {
        if (password == null) {
            password = '';
        }
        if (wordlist == null) {
            wordlist = LangEn.wordlist();
        }
        const mnemonic = Mnemonic.fromEntropy(randomBytes(16), password, wordlist);
        return this.createInstance(mnemonic);
    }
    /**
     * Creates an HD wallet from a phrase.
     *
     * @param {new (root: HDNodeWallet) => T} this - The constructor of the HD wallet.
     * @param {string} phrase - The phrase.
     * @param {string} [password] - The password.
     * @param {Wordlist} [wordlist] - The wordlist.
     * @returns {T} The created instance.
     */
    static fromPhrase(phrase, password, wordlist) {
        if (password == null) {
            password = '';
        }
        if (wordlist == null) {
            wordlist = LangEn.wordlist();
        }
        const mnemonic = Mnemonic.fromPhrase(phrase, password, wordlist);
        return this.createInstance(mnemonic);
    }
    /**
     * Connects the wallet to a provider.
     *
     * @param {Provider} provider - The provider.
     */
    connect(provider) {
        this.provider = provider;
    }
    /**
     * Validates the zone.
     *
     * @param {Zone} zone - The zone.
     * @throws {Error} If the zone is invalid.
     */
    validateZone(zone) {
        if (!Object.values(Zone).includes(zone)) {
            throw new Error(`Invalid zone: ${zone}`);
        }
    }
    /**
     * Serializes the HD wallet state into a format suitable for storage or transmission.
     *
     * @returns {SerializedHDWallet} An object representing the serialized state of the HD wallet, including version,
     *   mnemonic phrase, coin type, and addresses.
     */
    serialize() {
        return {
            version: this.constructor._version,
            phrase: this._root.mnemonic.phrase,
            coinType: this.coinType(),
        };
    }
    /**
     * Deserializes a serialized HD wallet object and reconstructs the wallet instance. This method must be implemented
     * in the subclass.
     *
     * @param {SerializedHDWallet} _serialized - The serialized object representing the state of an HD wallet.
     * @returns {AbstractHDWallet} An instance of AbstractHDWallet.
     * @throws {Error} This method must be implemented in the subclass.
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    static async deserialize(_serialized) {
        throw new Error('deserialize method must be implemented in the subclass');
    }
    /**
     * Validates an address info object, including basic properties and address derivation.
     *
     * @param {T} info - The address info to validate
     * @throws {Error} If validation fails
     * @protected
     */
    validateAddressInfo(info) {
        // Basic property validation
        this.validateBaseAddressInfo(info);
        // Validate address derivation
        this.validateAddressDerivation(info);
        // Allow subclasses to add their own validation
        this.validateExtendedProperties(info);
    }
    /**
     * Validates the NeuteredAddressInfo object.
     *
     * @param {NeuteredAddressInfo} info - The NeuteredAddressInfo object to be validated.
     * @throws {Error} If the NeuteredAddressInfo object is invalid.
     * @protected
     */
    validateBaseAddressInfo(info) {
        if (!/^(0x)?[0-9a-fA-F]{40}$/.test(info.address)) {
            throw new Error(`Invalid NeuteredAddressInfo: address must be a 40-character hexadecimal string prefixed with 0x: ${info.address}`);
        }
        if (!/^0x[0-9a-fA-F]{66}$/.test(info.pubKey)) {
            throw new Error(`Invalid NeuteredAddressInfo: pubKey must be a 66-character hexadecimal string prefixed with 0x: ${info.pubKey}`);
        }
        if (!Number.isInteger(info.account) || info.account < 0) {
            throw new Error(`Invalid NeuteredAddressInfo: account must be a non-negative integer: ${info.account}`);
        }
        if (!Number.isInteger(info.index) || info.index < 0) {
            throw new Error(`Invalid NeuteredAddressInfo: index must be a non-negative integer: ${info.index}`);
        }
        if (!Object.values(Zone).includes(info.zone)) {
            throw new Error(`Invalid NeuteredAddressInfo: zone '${info.zone}' is not a valid Zone`);
        }
    }
    /**
     * Hook for subclasses to add their own validation logic. Base implementation does nothing.
     *
     * @param {T} _info - The address info to validate
     * @protected
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    validateExtendedProperties(_info) {
        // Base implementation does nothing
    }
    /**
     * Validates the version and coinType of the serialized wallet.
     *
     * @param {SerializedHDWallet} serialized - The serialized wallet data to be validated.
     * @throws {Error} If the version or coinType of the serialized wallet does not match the expected values.
     * @protected
     * @static
     */
    static validateSerializedWallet(serialized) {
        if (serialized.version !== this._version) {
            throw new Error(`Invalid version ${serialized.version} for wallet (expected ${this._version})`);
        }
        if (serialized.coinType !== this._coinType) {
            throw new Error(`Invalid coinType ${serialized.coinType} for wallet (expected ${this._coinType})`);
        }
    }
}
export { AbstractHDWallet };
