// quais/lib/commonjs/wallet/bip44/bip44.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BIP44 = exports.MAX_ADDRESS_DERIVATION_ATTEMPTS = exports.HARDENED_OFFSET = void 0;
const utils_js_1 = require("../utils.js");
const index_js_1 = require("../../utils/index.js");
const bip32_js_1 = require("../bip32/bip32.js");
const secp256k1_1 = __importDefault(require("@bitcoinerlab/secp256k1"));
const index_js_2 = require("../../address/index.js");
const buffer_1 = require("buffer");
/**
 * Constant to represent the hardened offset for BIP44 derivation.
 */
exports.HARDENED_OFFSET = 2 ** 31;
/**
 * Constant to represent the maximum attempt to derive an address.
 */
exports.MAX_ADDRESS_DERIVATION_ATTEMPTS = 10000000;
/**
 * Constant to represent the maximum non-hardened index.
 */
const MAX_NON_HARDENED_INDEX = 0x7fffffff; // 2^31 - 1
class BIP44 {
    _coinType;
    /**
     * Root node of the HD wallet.
     */
    _root;
    constructor(root, coinType) {
        this._root = root;
        this._coinType = coinType;
    }
    /**
     * Returns the coin type of the wallet.
     *
     * @returns {AllowedCoinType} The coin type.
     */
    coinType() {
        return this._coinType;
    }
    /**
     * Returns the extended public key of the root node of the HD wallet.
     *
     * @returns {string} The extended public key.
     */
    get xPub() {
        return this._root.extendedKey;
    }
    /**
     * Returns the parent path for a given coin type.
     */
    parentPath() {
        if (!this._coinType) {
            throw new Error('Coin type not set');
        }
        return `m/44'/${this._coinType}'`;
    }
    /**
     * Gets the BIP44 change node for a given account and change flag.
     *
     * @param {number} account - The account number.
     * @param {boolean} change - Whether to get the change node.
     * @returns {HDNodeWallet} The change node.
     */
    _getChangeNode(account, change) {
        const changeIndex = change ? 1 : 0;
        return this._root.deriveChild(account + exports.HARDENED_OFFSET).deriveChild(changeIndex);
    }
    /**
     * Gets the BIP44 address node for a given account, change flag, and address index.
     *
     * @param {number} account - The account number.
     * @param {boolean} change - Whether to get the change node.
     * @param {number} addressIndex - The address index.
     * @returns {HDNodeWallet} The address node.
     */
    _getAddressNode(account, change, addressIndex) {
        return this._getChangeNode(account, change).deriveChild(addressIndex);
    }
    /**
     * Derives the next valid address node for a specified account, starting index, and zone. The method ensures the
     * derived address belongs to the correct shard and ledger, as defined by the Quai blockchain specifications.
     *
     * @param {number} account - The account number from which to derive the address node.
     * @param {number} startingIndex - The index from which to start deriving addresses.
     * @param {Zone} zone - The zone (shard) for which the address should be valid.
     * @param {boolean} [isChange=false] - Whether to derive a change address. Default is `false`
     * @returns {HDNodeWallet} - The derived HD node wallet containing a valid address for the specified zone.
     * @throws {Error} If a valid address for the specified zone cannot be derived within the allowed attempts.
     */
    deriveNextAddressNode(coinType, account, startingIndex, zone, isChange = false) {
        const changeNode = this._getChangeNode(account, isChange);
        let addrIndex = startingIndex;
        let addressNode;
        for (let attempts = 0; attempts < exports.MAX_ADDRESS_DERIVATION_ATTEMPTS; attempts++) {
            addressNode = changeNode.deriveChild(addrIndex++);
            if ((0, utils_js_1.isValidAddressForZone)(this.coinType(), addressNode.address, zone)) {
                return addressNode;
            }
        }
        throw new Error(`Failed to derive a valid address for the zone ${zone} after ${exports.MAX_ADDRESS_DERIVATION_ATTEMPTS} attempts.`);
    }
    /**
     * Derives the next valid address node for a specified account, starting index, and zone using React Native fast
     * crypto. This async version uses react-native-fast-crypto for secp256k1 operations. The method ensures the derived
     * address belongs to the correct shard and ledger, as defined by the Quai blockchain specifications.
     *
     * @param {number} account - The account number from which to derive the address node.
     * @param {number} startingIndex - The index from which to start deriving addresses.
     * @param {Zone} zone - The zone (shard) for which the address should be valid.
     * @param {boolean} [isChange=false] - Whether to derive a change address. Default is `false`
     * @returns {Promise<HDNodeWallet>} - The derived HD node wallet containing a valid address for the specified zone.
     * @throws {Error} If a valid address for the specified zone cannot be derived within the allowed attempts.
     */
    async deriveNextAddressNodeReactNativeAsync(account, startingIndex, zone, isChange = false) {
        const changeNode = this._getChangeNode(account, isChange);
        let addrIndex = startingIndex;
        let addressNode;
        for (let attempts = 0; attempts < exports.MAX_ADDRESS_DERIVATION_ATTEMPTS; attempts++) {
            addressNode = await changeNode.deriveChildReactNative(addrIndex++);
            if ((0, utils_js_1.isValidAddressForZone)(this.coinType(), addressNode.address, zone)) {
                return addressNode;
            }
        }
        throw new Error(`Failed to derive a valid address for the zone ${zone} after ${exports.MAX_ADDRESS_DERIVATION_ATTEMPTS} attempts.`);
    }
    /**
     * Performs non-hardened child derivation from a public key and chain code. This allows deriving child addresses at
     * CPU speed without private key access.
     *
     * @param {string} publicKey - The compressed public key (33 bytes as hex string)
     * @param {string} chainCode - The chain code (32 bytes as hex string)
     * @param {number} account - The account index (already hardened, e.g., 0' becomes 0)
     * @param {number} change - The change index (0 for external, 1 for internal)
     * @param {number} addressIndex - The address index
     * @returns {Object} Object containing derived public key, chain code, and address
     */
    static deriveChildFromPublic(publicKey, chainCode, account = 0, change = 0, addressIndex = 0) {
        // Validate indices are within non-hardened range (0 to 2^31 - 1)
        if (change < 0 || change > MAX_NON_HARDENED_INDEX) {
            throw new Error(`Invalid change index: ${change}. Must be between 0 and ${MAX_NON_HARDENED_INDEX}`);
        }
        if (addressIndex < 0 || addressIndex > MAX_NON_HARDENED_INDEX) {
            throw new Error(`Invalid address index: ${addressIndex}. Must be between 0 and ${MAX_NON_HARDENED_INDEX}`);
        }
        // Create BIP32 factory with ecc
        const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_1.default);
        // Create a BIP32 node from the public key and chain code
        const publicKeyBytes = (0, index_js_1.getBytes)(publicKey);
        const chainCodeBytes = (0, index_js_1.getBytes)(chainCode);
        // Validate public key length (33 bytes for compressed)
        if (publicKeyBytes.length !== 33) {
            throw new Error(`Invalid public key length: ${publicKeyBytes.length}. Expected 33 bytes for compressed public key`);
        }
        // Validate chain code length (must be exactly 32 bytes)
        if (chainCodeBytes.length !== 32) {
            throw new Error(`Invalid chain code length: ${chainCodeBytes.length}. Expected exactly 32 bytes`);
        }
        // Create node from public key
        let node = bip32.fromPublicKey(publicKeyBytes, chainCodeBytes);
        // Derive non-hardened path: change/addressIndex
        // Note: account is already hardened from Ledger, so we just derive change/addressIndex
        node = node.derive(change);
        node = node.derive(addressIndex);
        // Get the derived public key and compute the address
        const derivedPublicKey = node.publicKey;
        const derivedAddress = (0, index_js_2.computeAddress)('0x' + buffer_1.Buffer.from(derivedPublicKey).toString('hex'));
        const path = `${account}'/${change}/${addressIndex}`;
        return {
            publicKey: '0x' + buffer_1.Buffer.from(derivedPublicKey).toString('hex'),
            chainCode: '0x' + buffer_1.Buffer.from(node.chainCode).toString('hex'),
            address: derivedAddress,
            path: path,
        };
    }
}
exports.BIP44 = BIP44;
